[Do not review] Top of tracing per-param FSDP stack - to be broken up

yf225 opened the pull request:
Taking over the FSDP tracing work from https://github.com/pytorch/pytorch/pull/115410. Not all changes are copied from there - only the ones needed for tracing per-param FSDP.


**Dynamo / AOTAutograd code changes:**
1. queue_callback support (new)
2. Guard fix in generic_jump (new)
3. Relax constraint in _mirror_autograd_meta_to to allow requires_grad=True leaf tensor (already PR’ed: https://github.com/pytorch/pytorch/pull/120349) (new)
4. submodule registration on FSDP managed module (maybe hacky, need more thoughts) (copied from Voz's PR)
5. [TODO] Allow `resize_` and no-grad `_foreach_copy_` without issuing clone at beginning of graph.
6. In lieu of functionalization changes above, use custom FX passes to "re-inplace" the inplace ops in the FWD graph and CA graph, to 1) de-duplicate between saved tensors and hook closed-over tensors, 2) reduce `as_strided_scatter / copy_` etc ops and simplify the graph

**ppFSDP code changes:**
1. Move all pre-forward / post-forward / pre-backward / post-backward hooks to global scope in ppFSDP code (known requirement by compiled autograd hook support).
2. Under compile, keep unsharded_param as leaf tensor, and audit all uses of .all_gather_output and replace them with ._unsharded_param (particularly copy_out), so that `._unsharded_param` is graph input and `.all_gather_output` is *not* graph input. This is to avoid "pass multiple views (some of them undifferentiated) of the same tensor to the graph, and mutate some of them" problem which doesn't play well with AOTAutograd.
3. [Can improve] Gate all cuda stream and event usage with is_torchdynamo_compiling() in ppFSDP code. (known necessary change)
5. [Can improve] Under compile, use DTensor.from_local instead of _from_local_no_grad because DTensor() constructor is not supported by Dynamo yet. (new)
7. [Can improve] Under compile, use register_hook to call pre_backward and mimic multi_grad_hook "any" mode behavior. (new)
8. [Can improve] Under compile, ignore RegisterPostBackwardFunction, instead rely on _root_post_backward_final_callback to call the param_group.post_backward(), and rely on Inductor comm reordering to optimally decide when to issue prefetch and reshard.
9. [Hack] t.grad_fn is not supported by Dynamo yet, so we set grad_fns = [] and unconditionally do unshard in _prefetch_unshard() under compile. (new)
10. [Hack] Skip `self._fsdp_param_group.post_forward_mesh_info = None` under compile, because it seems to prevent `.resize_(0)` from issuing at end of forward graph under compile. (new)

**Change to how we run ppFSDP:**
1. [Can improve] Because Dynamo is not able to trace `_lazy_init` of ppFSDP yet, we have to run ppFSDP eager mode once to do the lazy initialization, and then trace ppFSDP. (new)

**If eager ppFSDP unit test fails:**
1. Look for `[EAGER TEST FAIL]` comment in the code.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
try view instead of as_strided for unsharded_param construction

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
let .all_gather_output view into .unsharded instead, to see if can remove clone at beginning of graph

yf225 committed:
Revert "let .all_gather_output view into .unsharded instead, to see if can remove clone at beginning of graph"

This reverts commit 2b968f39612c1a0bfe1f8ae528c0b599b239e926.

yf225 committed:
update

yf225 committed:
Revert "update"

This reverts commit a599dbb865d1e284377008d13fc32312c393f497.

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

awgu commented on torch/distributed/_composable/fsdp/_fsdp_state.py:
```diff
@@ -106,10 +188,13 @@ def _lazy_init(self) -> None:
-            # the parameters would be freed and all-gathered immediately
-            self._fsdp_param_group.post_forward_mesh_info = None
+        # NOTE(yf225): I believe this causes issue during compile
+        # (post_forward hook thinks that `post_forward_mesh_info` is None thus not doing resharding and not doing .resize_(0). Is it expected eager behavior?)
```
This is expected behavior. We choose to not call `resize_(0)` on the root FSDP module's parameters after forward since they would be immediately `resize_()`-d back and all-gathered.

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -246,8 +246,9 @@ def init_all_gather_output(
         if self.all_gather_output.numel() > 0:
             return  # already initialized
         all_gather_output_size = torch.Size([all_gather_input_numel * world_size])
+        # NOTE(yf225): here we explicitly make fsdp_param.all_gather_output the leaf tensor (instead of unsharded_param being leaf tensor)
```
I want to be careful about this for the long term solution.

The reason we have a distinction between `all_gather_output` and `unsharded_param` is because for certain extensions, they could be different. For example, if we were to support sharding QLoRA parameters, then there could be multiple `all_gather_output`s, but they will get transformed into a single subclass that is the `unsharded_param`.

However, for the default `torch.Tensor` case, `unsharded_param` should share the same data as `all_gather_output` but have any padding removed.

yf225 replied:
Discussed with Andrew offline: making `fsdp_param.all_gather_output` (i.e. padded unsharded) the requires_grad=True leaf tensor and making `unsharded_param` (i.e. unpadded unsharded) a differentiable view into it is okay to do.

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
update

yf225 committed:
add unit test for UDO as input to HOP

yf225 committed:
fix bug

yf225 committed:
use multi_grad_hook instead of RegisterPostBackwardFunction

yf225 committed:
try to use post_accumulate_grad_hook to mimic multi_grad_hook all mode

yf225 committed:
update

yf225 committed:
try to use register_hook instead of post_acc_grad_hook

yf225 committed:
Revert "try to use register_hook instead of post_acc_grad_hook"

This reverts commit 9b4829ca6e49b2e5c8966d78197f7f1d64d0f7f2.

yf225 committed:
Revert "try to use post_accumulate_grad_hook to mimic multi_grad_hook all mode"

This reverts commit 93cc88adb39c2be6e63dbf0755876b80ba273bb1.

yf225 committed:
Revert "use multi_grad_hook instead of RegisterPostBackwardFunction"

This reverts commit 409a47857a866bcdfa3d733af9aa7a07e738198a.

yf225 committed:
disable RegisterPostBackwardFunction when compile

yf225 committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
fix inductor_ops.cpp bug

yf225 committed:
use out-place cat then copy_, instead of torch.cat(out=...), in foreach_reduce_scatter_copy_in

yf225 committed:
Add several FX passes to do resize at the right time

yf225 committed:
update test file

yf225 committed:
add several more FX passes to simplify FSDP graph and work around functionalization issue; Inductor backend works for FC->RELU->FC now

yf225 committed:
set self._fsdp_param_group.post_forward_mesh_info = None in eager

yf225 committed:
try remove .all_gather_output leaf changes, and gate DTensor.from_local changes

yf225 committed:
use multi_grad_hook for pre_backward in eager

yf225 committed:
fix eager pre_backward bug

yf225 committed:
avoid using .all_gather_output as graph input; fix all eager mode tests

yf225 committed:
fix bugs

yf225 committed:
add FX pass: use input as output node for copy_

yf225 committed:
comment

yf225 committed:
add FX pass: if_tensor_is_resized_to_0_immediately_after_inplace_copy_then_delete_the_copy

yf225 committed:
add one FX pass to CA graph

yf225 committed:
remove assert any_leaf == all_leaf change

yf225 committed:
update test to use aot_eager

yf225 committed:
comment

yf225 committed:
comment out unused pass

yf225 committed:
make inductor backend work

yf225 committed:
clean

yf225 committed:
update test

yf225 committed:
update test

yf225 committed:
fixed test so that peak mem measurement is correct

yf225 committed:
add more FX passes to simplify CA BWD graph

yf225 committed:
add functorch config.enable_fsdp_fx_passes

yf225 committed:
use additional FX pass to fix peak mem usage

yf225 committed:
make compile resize to full target size a constant, to make it consistent between eager and compile

yf225 committed:
add fully_shard(_reshard_after_forward_root=...) config to allow post-forward reshard for root module params

yf225 committed:
simplify test script

yf225 committed:
add nested fully_shard test case

yf225 committed:
work around .use_training_state() ctx mgr and _register_post_backward_hook

yf225 committed:
clean up logs

yf225 committed:
simplify test

yf225 committed:
update FX passes

yf225 committed:
[WIP DEBUG] add compile to ppFSDP unit tests

yf225 committed:
Merge branch 'main' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
[Brian PR 122433] AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases

yf225 committed:
[Brian PR 122434] functionalize storage resizing, minimal ppFSDP traceable forward

yf225 committed:
fix merge issues

yf225 committed:
fix merge issues

yf225 committed:
fix debug tool

yf225 committed:
update unit test

yf225 committed:
remove all unnecessary FX passes after Brian's PRs

yf225 committed:
update fsdp unit test

yf225 committed:
work around torch/_dynamo/utils.py nn_module_proxy issue

yf225 committed:
add fsdp_fx_passes.replace_empty_then_slice_then_compute_then_foreach_copy_then_slice_scatter_pattern FX pass

yf225 committed:
disable joint_graph_constant_folding FX pass because it causes aten.full in graph

yf225 committed:
update test file

yf225 committed:
improve replace_foreach_all_gather_copy_out_pattern FX pass

yf225 committed:
fix _fsdp_state_pre_backward_compile_only bug

yf225 committed:
fixed trace_rules issue, enable test_multi_forward_module unit test

yf225 committed:
try to enable test_train_parity_2d_mlp

yf225 committed:
improved FX passes

yf225 committed:
add graph op check to enabled FSDP unit tests

yf225 committed:
add comments for custom ATen ops to replace FX passes

yf225 committed:
avoid repeatedly calling torch.compile in unit tests

yf225 committed:
Yifu's PR #122834 to fix reduce_scatter reduce type

yf225 committed:
improve test script test_dynamo_fsdp.py

yf225 committed:
add debug util to post_grad.py

yf225 committed:
Add Horace PRs: #122474, #122444

yf225 committed:
WIP Add FX pass: raise_all_gather_to_overlap_with_prev_layer_compute

yf225 committed:
improve FX passes

yf225 committed:
use _chunk_cat, remove need for foreach_reduce_scatter_copy_in replacement FX pass

yf225 committed:
add Python custom op for foreach_all_gather and foreach_all_gather_copy_out, remove need for most FSDP FX passes

yf225 committed:
raise_all_gather_to_overlap_with_prev_layer_compute and sink_reduce_scatter_wait_to_end_of_graph FX passes

yf225 committed:
add decompose_all_gather_copy_in FX pass

yf225 committed:
improve debug util

yf225 committed:
fix bug; clean up

yf225 committed:
add functorch config to enable/disble view chain movement

yf225 committed:
reenable joint_graph_constant_folding

yf225 committed:
add AC test to test_dynamo_fsdp.py

yf225 committed:
update test_dynamo_fsdp.py

yf225 committed:
remove contiguous_view_as_strided custom op

yf225 committed:
change FSDP hooks to staticmethod

yf225 committed:
remove need to gate stream/event with is_compiling

yf225 committed:
Add support for instance method in hook

yf225 committed:
switch to use original stream/event funcs

yf225 committed:
Add support for streams created outside of compile region

yf225 committed:
check whether all_gather_work is AsyncCollectiveTensor before wait

yf225 committed:
clean up

yf225 committed:
re-enable RegisterPostBackwardFunction

yf225 committed:
clean

yf225 committed:
check dist.distributed_c10d.Work instead

xmfan committed:
[aot] always pass inputs to runtime_wrapper as list and add type annotations (#123630)

`runtime_wrapper` unpacking the arguments as a Tuple[arg] will prevent them from being freed within its scope. This is problematic if inductors wants to free those inputs, which could be activations in the compiled backwards case. This PR only changes the signature to pass as list, but does not clear it, keeping same refcount as before.

Also adding some mypy annotations. Ideally, instead of `Any`, I would want a type to describe single arg which seems to be usually Tensor or int.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/123630
Approved by: https://github.com/jansel, https://github.com/bdhirsh

yf225 committed:
passing UDO into RegisterPostBackwardFunction doesn't work yet

yf225 committed:
fix test for ac

yf225 committed:
comment out unnecessary passes

xmfan committed:
[aot] refactor runtime_wrapper's epilogue args access (#123674)

I want runtime_wrapper args to be stealable by call_func_at_runtime_with_args, since the args may contain activations which we don't want to hold alive in this scope.

The args to runtime_wrapper **should always be** from a list created within aot_autograd, so it **should always be** safe to steal them: https://github.com/pytorch/pytorch/blob/a4a49f77b8c45ea459263c2242ab391b3d0577f2/torch/_functorch/aot_autograd.py#L928-L932

There are some accesses after we execute the compiled_fn, but those index accesses are already inferred at compile time.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/123674
Approved by: https://github.com/jansel, https://github.com/bdhirsh
ghstack dependencies: #123630

yf225 committed:
SAC wip

xmfan committed:
[compiled autograd][dynamo] Make compiled graph take in boxed inputs (#122353)

### Context
In today's Dynamo, we lift all tensors encountered during tracing to be individual graph inputs, even when they were in a container.

And [Dynamo generates](https://github.com/pytorch/pytorch/blob/fdc281f2587f9a5a935de1f1368e7ad7ed0f9828/torch/_dynamo/codegen.py#L371) the runtime function's signature using the graph's graphargs.

This means that the generated function will have each grapharg as an argument, which is problematic if we want to free the inputs in inductor codegen. See [python function arguments are kept alive for the duration of the function call](https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670).

```python
# original code
def forward(inputs):
  a, b, c, d, e = inputs
  inputs.clear()
  out = a
  out += b
  del b  # frees memory
  out += c
  del c  # frees memory
  out += d
  del d  # frees memory
  out += e
  del e  # frees memory
  return out

# compiled code:
def forward(a, b, c, d, e):
  # b, c, d, e can't be freed before end of function
```

This isn't a concern when compiling forward because a, b, c, d, e are all from user code, and should be kept alive. But when compiling backwards, a, b, c, d, e may be intermediate results i.e. activations, that we DO want to clear ASAP to remain on par with eager peak memory.

### Solution

We have encountered similar memory problems in AOTAutograd before, where we adopted the boxed calling convention (wrapping to-be-freed objects in a list), adding list clearing to inductor codegen, and being careful about holding references to elements in the input list. We need to do something similar, but for inputs from the user program (compiled autograd fx graph in this case).

This PR support lists as graphargs/placeholder nodes. When tracing a list of tensors, we create a node for it, and pre-emptively initialize variable trackers for its elements before they are used in the user program. Subsequent uses of those variables will find hits in the lookup table `input_source_to_var`.

With the inputs as a list in the graph args, our compiled code can free inputs just like in the eager case.
```python
def forward(inputs):
  # a, b, c, d, e can be freed within the function now
```

Currently, AOT/Inductor flattens list input via [flatten_graph_inputs wrapper](https://github.com/pytorch/pytorch/blob/597f479643f82859307ece38971f1c8e7d657c80/torch/_inductor/compile_fx.py#L1454-L1478), which is why this PR's CI can be green. Additional changes are needed to its runtime wrapper, done in the next PR. The next step is to ensure that we are careful in forwarding the list to inductor codegen without holding additional references.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/122353
Approved by: https://github.com/jansel
ghstack dependencies: #123630, #123674

yf225 committed:
Merge branch 'main_56dd760' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_04092024_main

yf225 committed:
Simon CA mem fix stack https://github.com/pytorch/pytorch/pull/122535

yf225 committed:
Merge branch 'main_56dd760_simon_CA_mem_fix' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_04092024_main

yf225 committed:
fix merge issue

yf225 committed:
update test script

yf225 committed:
fix merge issue 2

xmfan committed:
[compiled autograd][dynamo] Codegen aliases to keep grad mutated tensors alive (#123359)

The current codegen is problematic if __compiled_fn_0 clears the inputs list, since we need it for assignment afterwards
```python
def forward(inputs):
    __compiled_fn_0 = ...  # The actual function needs to be provided
    graph_out_0 = __compiled_fn_0(inputs)  # clears inputs
    temp_list = []
    temp_list.append(graph_out_0[0])
    inputs[4].grad = graph_out_0[1]  # inputs is empty, index error
    inputs[7].grad = graph_out_0[2]
    inputs[8].grad = graph_out_0[3]
    inputs[9].grad = graph_out_0[3]
    del graph_out_0
    return temp_list
```

With this fix, we use aliases to keep the tensors alive
```python
def forward(inputs):
    __compiled_fn_0 = ...  # The actual function needs to be provided
    inputs_ref_1 = inputs[9]
    inputs_ref_2 = inputs[4]
    inputs_ref_3 = inputs[8]
    inputs_ref_4 = inputs[7]
    graph_out_0 = __compiled_fn_0(inputs)
    temp_list = []
    temp_list.append(graph_out_0[0])
    inputs_ref_2.grad = graph_out_0[1]
    inputs_ref_4.grad = graph_out_0[2]
    inputs_ref_3.grad = graph_out_0[3]
    inputs_ref_1.grad = graph_out_0[3]
    del graph_out_0
    return temp_list
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/123359
Approved by: https://github.com/jansel
ghstack dependencies: #123630, #123674, #122353

xmfan committed:
[compiled autograd][aot] Trim runtime refs for list inputs from dynamo (#122535)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/122535
Approved by: https://github.com/bdhirsh
ghstack dependencies: #123630, #123674, #122353, #123359

yf225 committed:
Revert "Simon CA mem fix stack https://github.com/pytorch/pytorch/pull/122535"

This reverts commit 77691819892408a5f194f2d5251678d718dabc9f.

yf225 committed:
wip

yf225 committed:
fix issues:

yf225 committed:
print ref counts

yf225 committed:
fix print refcount

yf225 committed:
updated AC test

yf225 committed:
clean up prints

yf225 committed:
update comment

yf225 committed:
Apply Simon's PRs: #124127, #124155

yf225 committed:
Memory fix: issue  for wait_tensor output

yf225 committed:
Memory fix: sink RS wait to before next RS

yf225 committed:
add debug util

yf225 committed:
fix merge bug

yf225 committed:
update test script to fix memory snapshot issues

yf225 committed:
add toy Transformer model to test_dynamo_fsdp.py

yf225 committed:
improve test script

yf225 committed:
improve test script

yf225 committed:
fix RS sink wait FX pass

yf225 committed:
move most passes to Inductor scheduler

yf225 committed:
fix FSDP2+SAC peak memory usage issue, by performing reordering pass in Inductor to release memory sooner

yf225 committed:
comment

yf225 committed:
add config flags

yf225 committed:
clean

yf225 committed:
update test

yf225 committed:
comment

yf225 committed:
comment

yf225 committed:
comment

yf225 committed:
rename mem pass config

yf225 committed:
rename

yf225 committed:
clean up

yf225 committed:
improve unit test

yf225 committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
remove old storage resize related code

yf225 committed:
Apply Brian's PR: #122434

yf225 committed:
fix bug

yf225 committed:
go back to test simple full_graph case

yf225 committed:
simple test passes

yf225 committed:
fix RNG higher-order op related bug

yf225 committed:
fix simple test

yf225 committed:
add remove_storage_resize_and_copy FX pass

yf225 committed:
fix bug

yf225 committed:
fix minor change in FSDP

yf225 committed:
add FX pass: reinplace_primal_copyout_from_allgather_output

yf225 committed:
update test

yf225 committed:
fix memory bug, llama now working with good peak memory

yf225 committed:
fix pass ordering

yf225 committed:
fix memory issue

yf225 committed:
also raise users that are fused nodes

yf225 committed:
fix memory usage and overlap

yf225 committed:
avoid need to run 1st iteration in eager

yf225 committed:
add graph break test

yf225 committed:
Revert "Apply Brian's PR: #122434"

This reverts commit 5e454115d890e12f232a9e443e7fcbcb156fdbfe.

yf225 committed:
Merge branch 'main' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
Inductor scheudler debug_str fix

yf225 committed:
clean up logs

yf225 committed:
remove test files

yf225 committed:
simplify DTensor from_local

yf225 committed:
experiment on queue_callback impl:

yf225 committed:
[Traceable FSDP][Compiled Autograd] Add queue_callback() support

yf225 committed:
fix

yf225 committed:
lint

yf225 committed:
clean up test script

yf225 committed:
more exp on queue_callback

yf225 committed:
more exp on queue_callback

yf225 committed:
clean up

yf225 committed:
clean up

yf225 committed:
clean

yf225 committed:
add CompiledAutogradEngine to external_utils.py

yf225 committed:
clean

yf225 committed:
remove node.py changes:

yf225 committed:
simplify

yf225 committed:
improve

yf225 committed:
lint

yf225 committed:
comment

yf225 committed:
Merge branch 'main' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
Merge branch 'main' of https://github.com/yf225/pytorch into compile_fsdp_queue_callback

yf225 committed:
Revert "more exp on queue_callback"

This reverts commit 53c4614884fde6afadea56823f28556db5cd704c.

yf225 committed:
Revert "clean up"

This reverts commit 2f540c4102c4abeb7f2fa71f79bcf4d8b08ed143.

yf225 committed:
Revert "experiment on queue_callback impl:"

This reverts commit b20084756467122b0ce6b8e1ce30f4aa5877c33c.

yf225 committed:
Merge branch 'compile_fsdp_queue_callback' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
fix queue_callback merge

yf225 committed:
work around out= lowering

yf225 committed:
simpler way to call 1st iter with eager

yf225 committed:
make graph breka test harder

yf225 committed:
remove multi_grad_hook usage

yf225 committed:
fix FSDP2+TP compile test

yf225 committed:
improve test script

yf225 committed:
support AGO and nn.Parameter creation within graph

yf225 committed:
removed eager/compile divergence for AG copy-out; aot_eager backend working, inductor pass need fix

yf225 committed:
improve test

yf225 committed:
Merge branch 'main' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 committed:
remove move_view_chain_to_bwd_graph flag

yf225 committed:
WIP trying to make +TP also work

yf225 committed:
turn off prints

yf225 committed:
simplify alloc_storage

yf225 committed:
simplify

yf225 committed:
Merge branch 'main_70d8bc2' of https://github.com/yf225/pytorch into per_param_FSDP_dynamo_tracing_based_on_02202024_main

yf225 commented:
Superseded by https://github.com/pytorch/pytorch/pull/127239.
